Include comments in config files generated with --create-config-file

SylviaWhittle opened the pull request:
Closes #536 

This PR changes the behaviour of generating config files using the command-line flag `--create-config-file`, so that it uses `shutil` to copy the `default_config.yaml` file directly. This is a trade-off on functionality, but it seems that users would prefer the convenience of having the comments in the file.

If it turns out that the users do not end up liking this change, it should be trivial to reverse it as it's a small change that does not affect any other code as far as I can tell.

ns-rse commented on topostats/run_topostats.py:
```diff
@@ -163,12 +164,12 @@ def main(args=None):
-            config_file=args.create_config_file,
-            header_message="Sample configuration file auto-generated",
-        )
+        default_config_path = pkg_resources.path(__package__, "default_config.yaml")
```
I understand why this is here but I think we can make things a little simpler by changing some code earlier on.

Currently on lines 146-150 we have...

```python
    if args.config_file is not None:
        config = read_yaml(args.config_file)
    else:
        default_config = pkg_resources.open_text(__package__, "default_config.yaml")
        config = yaml.safe_load(default_config.read())
```

...and so we have two lines that are very similar here as they both have `pkg_resources.open_text(__package__, "default_config.yaml")`.

If we change the logic of the earlier code we only need to have one line for setting the path to the `default_config.yaml`...

```python
    if args.config_file is None:
        default_config = pkg_resources.open_text(__package__, "default_config.yaml")
        config = yaml.safe_load(default_config.read())
    else:
        config = read_yaml(args.config_file)
```

...and the above line is no longer needed (although line 172 would need changing to use `default_config` rather than `default_config_path`).

SylviaWhittle replied:
I could be misunderstanding you. 
```python
pkg_resources.open_text(__package__, "default_config.yaml")
```
returns a string of the contents of the config file, which is not usable by `shutil.copy`, which takes a file system path. This could work if I manually saved to the file using `with open() as f`, is this what you intended?

ns-rse commented on topostats/run_topostats.py:
```diff
@@ -163,12 +164,12 @@ def main(args=None):
-        )
+        default_config_path = pkg_resources.path(__package__, "default_config.yaml")
+        if ".yaml" not in args.create_config_file and ".yml" not in args.create_config_file:
+            create_config_path = "./" + args.create_config_file + ".yaml"
```
Not tested this on M$-Win but would the hard coding of `./` cause any problems? I've a vague memory that the Windows Linux Shell does interpret/handle these correctly. If not perhaps using `pathlib.Path("./")` might mitigate the problem (if there is one!).

ns-rse reviewed (commented):
Thanks for getting this one sorted too @SylviaWhittle 

I've suggested a slight change to avoid duplicated lines simply by changing earlier logic.

I think its a bit of a shame to lose the file header which indicates when the sample config was generated but date/time isn't too reliable given people could be on different branches anyway and besides we ask for the config used at any given point to be included in Bug Reports anyway.

SylviaWhittle commented:
Thanks for the comments @ns-rse, I've just pushed a re-jig of it, where I use your suggestion above, and I create a function in `io.py`, `write_config_with_comments` to handle the file writing.

I had to add a line to ensure users are not running with `--create-config-file` and `--config`, as this would leave `default_config` undefined before its call.

I've also moved things about a bit with the documentation, so that we keep the `# Config file generated 2023-05-27 01:01:12` and documentation links. 
There were some 404 links in the code that I have removed, leaving only a link that works. This should also mean that it should be easy to have all documentation hints standardised and removed link duplication in the codebase

Let me know what you think and I'll add tests if you approve (don't worry I hadn't forgotten about adding tests). 👍

ns-rse commented on topostats/io.py:
```diff
@@ -67,17 +87,14 @@ def write_yaml(
     if header_message:
         config_yaml.yaml_set_start_comment(
-            f"{header_message} : {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n" + documentation_reference
+            f"{header_message} : {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n" + CONFIG_DOCUMENTATION_REFERENCE
```
Could use the `get_date_time()` function here as you do below.

ns-rse commented on topostats/io.py:
```diff
@@ -86,6 +103,32 @@ def write_yaml(
+        A name for the configuration file. Can have a ".yaml" on the end.
+    """
+
+    if ".yaml" not in filename and ".yml" not in filename:
```
Good to have this eventually caught :+1:

ns-rse commented on topostats/io.py:
```diff
@@ -86,6 +103,32 @@ def write_yaml(
             LOGGER.error(exception)
 
 
+def write_config_with_comments(config: str, filename: str = "config.yaml"):
```
Minor pedantic typehint for the return value.

```suggestion
def write_config_with_comments(config: str, filename: str = "config.yaml") -> None:
```

ns-rse commented on topostats/io.py:
```diff
@@ -86,6 +103,32 @@ def write_yaml(
+    """
+
+    if ".yaml" not in filename and ".yml" not in filename:
+        create_config_path = "./" + filename + ".yaml"
```
Might be overkill using `Path.cwd()` here and under the `else:` but could.

```suggestion
        create_config_path = Path.cwd() / f"{filename}.yaml"
```

ns-rse reviewed (approved):
Neat, I like this solution of using a string object rather than `shutil.copy` and that you've retained the file header at the same time.

Not sure its absolutely necessary to use `Path.cwd()` may be overkill. :shrug:

ns-rse reviewed (approved):
Looks good to go when the tests complete. :+1:
